src/StructTypes.jl

```diff
@@ -1089,7 +1102,7 @@ constructfrom(::ArrayType, ::Type{T}, obj) where {T <: Tuple} =
 
 function constructfrom(::AbstractType, ::Type{T}, obj::S) where {T, S}
     types = subtypes(T)
-    if length(types) == 1
+    if length(types) == 1 && !(types isa SubTypeClosure)
```

Comment:
So looking over packages that _use_ StructTypes, we have this codepath in several, so this is my only hesitation because I'm worried things will break, for example, in JSON3.jl, since it doesn't have this same check.

Could we instead define `Base.getindex(s::SubTypeClosure, i::Int) = first(s._subtypes)[2]`? I know it's weird, but it would help `SubTypeClosure` be just enough of a `NamedTuple` and "just work" w/ existing StructType dependents.

Reply:
I think this is a great call, I'll edit the PR. Definitely don't want to break things for downstream users.